update TransactionList to new events/design

mannynotfound opened the pull request:
## Description

- add retrieval of `uuid` when getting treasuries so we can use that to query events
- update `refreshTreasury` / `fetchTreasury` to do a single request that fetches `uuid` + `signers` + `threshold`
- remove old transactions detail modal, going to implement new one in follow up PR

## Demo / Test Result

![image](https://user-images.githubusercontent.com/15314629/191364123-5a621a00-a2d2-4b45-b665-291b4bdc5430.png)

mannynotfound commented on packages/client/src/components/TransactionList.js:
```diff
@@ -190,17 +112,27 @@ function TransactionList({ safeData, transactions = [{ foo: "bar" }] }) {
+  transactions.forEach(tx => {
+    // only show successes / rejections
+    const isTransfer = tx?.blockEventData?.actionView?.type.includes('Transfer');
+    const isRejection = tx?.flowEventId.includes('ActionDestroyed');
```
currently the design only shows successful transfers or rejected actions so filtering out anything else

mannynotfound commented on packages/client/src/components/TransactionList.js:
```diff
@@ -134,50 +28,78 @@ function TransactionDetails({ safeOwners, transaction, onClose }) {
+function getTransferType(intent) {
+  if (intent.includes("from the treasury")) {
+    return "Sent";
+  }
```
I don't love this but I'm not sure a better way to differentiate between in/out. It's not as easy as checking if `recipient === treasuryAddr` since the recipient could be the wallet address of the user that owns the treasury.

linxiao-li commented on packages/client/src/components/TransactionList.js:
```diff
@@ -1,78 +1,47 @@
-  if (!tokenTransfer) {
-    return "";
-  }
+const Column = ({ className = "", children, style = {} }) => (
```
Is this component purely for styling? What's the consideration of rendering by columns not rows?

mannynotfound replied:
was to reduce the boilerplate of `column is-flex is-align-items-center`. I guess we could rename `TransactionListItem` to `TransactionListRow` or simply `Row`

linxiao-li replied:
Yeah and looks like only 2 out of 5 columns need `is-flex is-align-items-center`

linxiao-li commented on packages/client/src/components/TransactionList.js:
```diff
@@ -1,78 +1,47 @@
 import TransactionStatusIcon from "./TransactionStatusIcon";
 
-const FALLBACK_RECIPIENT_NAME = "No name";
+const maxWidths = {
```
We should avoid rendering with hardcode pixel values unless absolutely necessary. Anyway we can solve the problem with flex-box or bulma columns?

mannynotfound replied:
its necessary to define the widths here so that `intent` (Info column) can be take up all available width as the strings can be quite long.

linxiao-li replied:
Would using flex-grow help?

linxiao-li commented on packages/client/src/components/TransactionList.js:
```diff
@@ -1,78 +1,47 @@
+  const friendlyNameMap = {
+    "FlowToken": "FLOW",
   };
+  const newWords = intent.split(" ").map(word => {
```
Can we move this to the utils?

mannynotfound replied:
I think utils should be reserved for functions that are used in several places

linxiao-li replied:
The function is to display the intent in a more friendly way, I can see this could be used in other places.

linxiao-li commented on packages/client/src/components/TransactionList.js:
```diff
@@ -1,78 +1,47 @@
+  return newWords.join(" ");
+}
 
+const TransactionDetails = ({ transaction, onClose }) => {
```
Can we move this into a separate file?

mannynotfound replied:
I will be in a follow up PR! I have a ticket for just the detail modal

linxiao-li commented on packages/client/src/components/TransactionList.js:
```diff
@@ -132,52 +62,53 @@ function TransactionDetails({ safeOwners, transaction, onClose }) {
+};
 
-function TransactionListItem({
+const TransactionListItem = ({
```
Can we move to a separate file?

mannynotfound replied:
this file is ~150 lines of code which is pretty average for our codebase, I think its fine to co-locate these as they are very coupled

linxiao-li replied:
I remember having a similar discussion before, it is true that those two components are very closely coupled, the problem is average length of file and how closely components are coupled are subjective, not as easy to execute on the simple rule of one component per file. As the codebase grows bigger and more complex, sometimes we need these cookie cutter rules to keep the codebase maintainable. @amyliu6 What's your opinion?

mannynotfound replied:
I would vote to follow CAST's eslint rules so we have parity across the organization: https://github.com/DapperCollectives/CAST/blob/main/frontend/packages/client/package.json#L80-L85

linxiao-li replied:
That's perfect! I'll make sure to add this rule to our eslint config

linxiao-li commented on packages/client/src/components/TransactionList.js:
```diff
@@ -132,52 +62,53 @@ function TransactionDetails({ safeOwners, transaction, onClose }) {
-          onClick={() => onView(transaction)}
-        >
-          View
+    <div className={classes}>
```
Seems the container div is missing "columns" class

linxiao-li commented on packages/client/src/components/TransactionList.js:
```diff
@@ -132,52 +62,53 @@ function TransactionDetails({ safeOwners, transaction, onClose }) {
-          View
+    <div className={classes}>
+      <Column style={{ maxWidth: maxWidths.index }}>
+        {String(displayIndex).padStart(2, "0")}{' '}
```
Can you help me understand what this part does?

mannynotfound replied:
this code was already here but as I understand it adds a zero to the beginning of the number until it reaches the desired length. This is so we show `01`, `02`, etc instead of `1`, `2,`

linxiao-li commented on packages/client/src/components/TransactionList.js:
```diff
@@ -190,17 +121,35 @@ function TransactionList({ safeData, transactions = [{ foo: "bar" }] }) {
+
+  return (
+    <div className="border-light rounded-sm">
+      <div className="is-flex has-text-grey border-light-bottom is-hidden-mobile">
```
missing columns class?

mannynotfound replied:
so `columns` actually has all this extra baggage that makes it easier to just not use it in most cases... i agree its weird but adding it here means I have to do a bunch of overrides as well lol

![image](https://user-images.githubusercontent.com/15314629/191599074-49df8842-0823-4850-9c63-1b83d8d6a88e.png)
![image](https://user-images.githubusercontent.com/15314629/191599120-c150406e-df61-4be4-84e6-370760b870bc.png)

linxiao-li replied:
I see, would `columns is-gapless` help solving the problem?

mannynotfound replied:
that adds this and a few other styles that are troublesome 

![image](https://user-images.githubusercontent.com/15314629/191608025-e222ba6f-b9ae-4be9-afdc-7e1ce7f97d80.png)

The reason for not using bulma `columns` is because it assumes things of equal width or in neat little fractions like 1/3 etc but really this is a table where most of the columns are always going to be same width except for the variable string "Info" column.

linxiao-li commented on packages/client/src/components/TransactionStatusIcon.js:
```diff
@@ -3,17 +3,17 @@ import React from "react";
 const STATUS_TO_BACKGROUND_COLOR = {
   confirmed: "#18AA01",
   pending: "#FF8A00",
+  rejected: "#CD0B0B",
```
@KsenijaZivkovic added a status Icon and colors in her pr, https://github.com/DapperCollectives/VESSEL/pull/108
I think this file can be deprecated now.

mannynotfound replied:
would rather update this in a follow up PR after #108 is merged to not be dependent on it, or if it gets merged before this PR does I can update

linxiao-li commented on packages/client/src/hooks/useTreasuryTransactions.js:
```diff
@@ -6,34 +6,17 @@ const NETWORK = process.env.REACT_APP_FLOW_ENV === "testnet" ? "test" : "main";
 const API_URL = `https://prod-${NETWORK}-net-dashboard-api.azurewebsites.net/api`;
 const FETCH_URL = `${API_URL}/company/${COMPANY_ID}/search?projectId=${PROJECT_ID}`;
```
I still don't think it's a good idea to have configs hardcoded in the source code, I'll work on implementing a solution for the configs https://linear.app/dappercollectives/issue/VSL-214

mannynotfound replied:
we could make `REACT_APP_GRAFFLE_API_URL_TESTNET` and `REACT_APP_GRAFFLE_API_URL_MAINNET` and then switch which one we use based on `NETWORK` but its just more code and this shouldn't need to change / be configured that often

linxiao-li replied:
I'm thinking more like a `useConfig` hook, to hide all these implementation details of string concatenation

linxiao-li reviewed (commented):
Looks like SafeTransaction and TransactionHistory is identical, any way we can combine and reuse?  Suggest breaking down TransactionList into separate files, and group TransactionHistory, TransactionList, TransactionDetails, TransactionItem into a folder to make files more organized.

mannynotfound commented:
@amyliu6 updated to actual latest figma designs!

![image](https://user-images.githubusercontent.com/15314629/191605837-10105902-c844-40a3-b1a5-7613dff4498e.png)

@linxiao-li renamed `TransactionList` => `TransactionTable` since that is more of a table and renamed `TransactionListItem` to `Row` to make more sense with `Column`. Also refactored `SafeTransactions`/`TransactionHistory` such that now we only have the code for fetching treasury events in `SafeTransactions` and `TransactionHistory` just imports `SafeTransactions`

KsenijaZivkovic commented on packages/client/src/components/TransactionStatusIcon.js:
```diff
@@ -3,17 +3,17 @@ import React from "react";
 const STATUS_TO_BACKGROUND_COLOR = {
   confirmed: "#18AA01",
   pending: "#FF8A00",
+  rejected: "#CD0B0B",
```
Can we use here the colors defined in App.sass instead?
We have success, warning and danger classes for it

KsenijaZivkovic commented on packages/client/src/components/TransactionTable.js:
```diff
@@ -0,0 +1,134 @@
+  const transactionsToShow = transactions.filter((tx) => tx?.blockEventData?.actionView?.intent);
+
+  return (
+    <div className={`border-light rounded-sm ${className}`}>
```
We can use the class .table-border to add to rows border light bottom

KsenijaZivkovic commented on packages/client/src/components/TransactionTable.js:
```diff
@@ -0,0 +1,134 @@
+const Row = ({ transaction, displayIndex, isLastInList, onView }) => {
+  const classes = [
+    "py-4 is-flex is-align-items-center is-justify-content-space-between",
+    isLastInList ? "" : "border-light-bottom",
```
We don't need the prop isLastInList, we can use class .table-border for parent element

mannynotfound commented on packages/client/src/components/SafeHome.js:
```diff
@@ -3,24 +3,11 @@ import SafeOverview from "./SafeOverview";
-    />
-    <TransactionHistory safeData={safeData} address={address} />
+    <SafeOverview {...props} />
+    <ThingsToDo {...props} />
```
`ThingsToDo` was missing `onReject` and `onApprove` so wasnt able to take those actions on items. since `SafeHome` is just a wrapper component that bundles other components to render, lets just spread all props to its children and children can decide what to use

linxiao-li replied:
Pasting it here again in case you missed it:
I would suggest against passing props to all children
1. It creates redundant dependencies, causing components to be dependent and to re-render on irrelevant state changes.
2. It hurts readability, in order to figure out what's been passed to the component, another developer has to to a lot of tracking and code tracing.

onReject and onApprove is missing because of a bad merge conflict, creating ways to accommodate it will lead to more error in the future

mannynotfound replied:
just trying to reduce boilerplate and places that need to be updated, but can update to xenias solution

mannynotfound commented:
@linxiao-li since xenia's PR merge I moved from `getFriendlyIntent` => `utils.formatActionString` and `TransactionStatusIcon` => `Svg name="Status"` 🎉 also just removed `column` and added the styles from the column I needed

@amyliu6 here's what mobile looks like, the "header" goes away and some of the columns are also hidden

![image](https://user-images.githubusercontent.com/15314629/192328479-99f11456-d638-4f02-85bf-607b3c57386d.png)

KsenijaZivkovic commented:
Looks good! One small thing: border seems to shrink when we emulate this on the mobile view.
<img width="197" alt="Screen Shot 2022-09-26 at 11 01 51 AM" src="https://user-images.githubusercontent.com/20346621/192348001-d23e65c6-abe1-44aa-b92e-14c37f628aef.png">

linxiao-li commented on packages/client/src/pages/Safe.js:
```diff
@@ -62,6 +62,18 @@ const ReceiveTokens = ({ name, address }) => {
   );
 };
 
+const EmptyTransactions = () => (
```
We have a `library/EmptyTableWithCTA` component you can modify and use

mannynotfound commented:
> Looks good! One small thing: border seems to shrink when we emulate this on the mobile view. <img alt="Screen Shot 2022-09-26 at 11 01 51 AM" width="197" src="https://user-images.githubusercontent.com/20346621/192348001-d23e65c6-abe1-44aa-b92e-14c37f628aef.png">

seems like this is coming from an unrelated style - we should probably fix this in a separate PR as it affects a lot of things
![image](https://user-images.githubusercontent.com/15314629/192365256-a9ed47d8-b785-42dd-baf1-bdc3a787d77d.png)

linxiao-li commented on packages/client/src/App.sass:
```diff
@@ -412,6 +412,9 @@ input:focus-visible
 .flex-3
   flex: 3 0 0
+.flex-9
+  flex: 9 0 0
```
Should probably set to flex: 9 9 9 (or flex:9 in short) if we want this column take 9 spaces on all screen sizes
